docs/src/main/asciidoc/writing-extensions.adoc

```diff
@@ -2025,16 +2025,103 @@ public final class MyExtProcessor {
         resource.produce(new NativeImageResourceBuildItem("/security/runtime.keys")); //<1>
 
         resource.produce(new NativeImageResourceBuildItem(
-                "META-INF/services/" + io.quarkus.SomeService.class.getName())); //<2>
+                "META-INF/my-descriptor.xml")); //<2>
 
         resourceBundle.produce(new NativeImageResourceBuildItem("javax.xml.bind.Messages")); //<3>
     }
 }
 ----
 <1> Indicate that the /security/runtime.keys classpath resource should be copied into native image.
-<2> Indicate that the `io.quarkus.SomeService` ServiceLoader resource should be copied into native image.
+<2> Indicate that the `META-INF/my-descriptor.xml` resource should be copied into native image
 <3> Indicate that the "javax.xml.bind.Messages" resource bundle should be copied into native image.
 
+==== Service files
+
+If you are using `META-INF/services` files you need to register the files as resources so that your native image can find them,
+but you also need to register each listed class for reflection so they can be instantiated or inspected at run-time:
+
+[source,java]
+----
+public final class MyExtProcessor {
+
+    @BuildStep
+    void registerNativeImageReources(BuildProducer<NativeImageResourceBuildItem> resource,
+                                     BuildProducer<ReflectiveClassBuildItem> reflectionClasses) {
+        String service = "META-INF/services/" + io.quarkus.SomeService.class.getName();
+        
+        // register the service file so it is visible in native-image
+        resource.produce(new NativeImageResourceBuildItem(service));
+        
+        // register every listed implementation class so they can be inspected/instantiated 
+        // in native-image at run-time
+        Set<String> implementations = 
+            ServiceUtil.classNamesNamedIn(Thread.currentThread().getContextClassLoader(),
+                                          service);
+        reflectionClasses.produce(
+            new ReflectiveClassBuildItem(true, true, implementations.toArray(new String[0])));
```

Comment:
What about `ServiceProviderBuildItem` ? I think this would simplify most of what you did there.

Reply:
HAH. Well, it's brand new (three weeks) and undocumented, but actually worse: there's a competing implementation of the same name in the extension author guide, which doesn't do the same at all: it aggregates all services of the same name into a single service file.

This is VERY confusing. How can we resolve this?

Reply:
The newness is only because of the package rename, the two that are there are exactly the same.

Reply:
The one that should be used is the one in the `.nativeimage.` package. Otherwise, it has been there since before the shamrock renaming in the `substrate` package.

The `substrate` one is deprecated and deprecation is handled via producing the `.nativeimage.` ones.

Reply:
OK, so it was just not documented in the guide and we still have a class with the same name in the guide. I'll rename the guide class then.

Reply:
OK wait. I don't understand this build item. It appears to do two things:

- register the service file as a resource
- register the classes for reflection

But it's invoked mostly while specifying specific implementation classes, and only once by reading the actual service file to extract the implementation lists. It does not appear to write the service file with the list of resources it's given (unlike the version with the same name in the guide). So how does it actually work? If all the callers specify implementation classes that do not actually end up in the service files, how can the service loader see them?

Or are they _also_ manually put in the service file and the list of parameters happens to match what's in the existing file?

That's a pretty important thing to clear up @stuartwdouglas please?

Reply:
@stuartwdouglas I can document the fact that this build item depends on the list of implementations passed as params matching existing service files, but I need your confirmation that it does this to unstuck my PR.

Reply:
It does. Really we should just have a simple way of enabling all services by reading the service files from the deployment, but I guess it is not a major issue as nobody has really complained yet.

Reply:
OK thanks, I'll document this then.